pointmatcher/DataPointsFilters/utils/octree.hpp

```diff
@@ -0,0 +1,609 @@
+			XYZ{-0.5, +0.5, +0.5},
+			XYZ{+0.5, +0.5, +0.5}
+		};
+	//Check maxData count
+	if((bb.radius*2.0 <= maxSizeByNode) or (datas.size() <= 1))
+	{
+		//insert data
+		data.insert(data.end(), 
+			std::make_move_iterator(datas.begin()), make_move_iterator(datas.end()));
+		return (isLeaf());
+	}
+	
+	//Assign bounding box
+	this->bb.center = bb.center;
+	this->bb.radius = bb.radius;
+	
+	//Split datas
+	const std::size_t nbData = datas.size();
+	
+	DataContainer sDatas[8];
+	for(size_t i=0; i<8; ++i)
+		sDatas[i].reserve(nbData);
+		
+	for(auto&& d : datas)
+	{
+		//FIXME: Should be a generic conversion from DataPoint considering Data to XYZ
+		(sDatas[idx( TO_XYZ(pts,d) )]).emplace_back(d);
+	}
+	
+	for(size_t i=0; i<8; ++i)
+		sDatas[i].shrink_to_fit();
+	
+	//Compute new bounding boxes
+	BoundingBox boxes[8];
+	const T half_radius = this->bb.radius * 0.5;
+	for(size_t i=0; i<8; ++i)
+	{
+		const XYZ offset = offsetTable[i] * this->bb.radius;
+		boxes[i].radius = half_radius;
+		boxes[i].center = this->bb.center + offset;
+	}
+	
+	//For each child build recursively
+	bool ret = true;
+	for(size_t i=0; i<8; ++i)
+	{
+		octants[i] = new Octree<T>();
+		octants[i]->depth = this->depth+1;	
+		ret = ret and octants[i]->build(pts, std::move(sDatas[i]), std::move(boxes[i]), maxSizeByNode);		
+		//Assign parent
+		octants[i]->parent = this;
+	}
+
+	return (!isLeaf() and ret);
+}
+
+template< typename T >
+bool Octree<T>::build_par(const DP& pts, DataContainer&& datas, BoundingBox && bb, T maxSizeByNode)
+{
+	static XYZ offsetTable[8] =
+		{
+			XYZ{-0.5, -0.5, -0.5},
+			XYZ{+0.5, -0.5, -0.5},
+			XYZ{-0.5, -0.5, +0.5},
+			XYZ{+0.5, -0.5, +0.5},
+			XYZ{-0.5, +0.5, -0.5},
+			XYZ{+0.5, +0.5, -0.5},
+			XYZ{-0.5, +0.5, +0.5},
+			XYZ{+0.5, +0.5, +0.5}
+		};
+	
+	//Check maxData count
+	if((bb.radius*2.0 <= maxSizeByNode) or (datas.size() <= 1))
+	{			
+		//insert data
+		data.insert(data.end(), 
+			std::make_move_iterator(datas.begin()), make_move_iterator(datas.end()));
+		return (isLeaf());
+	}
+	
+	//Assign bounding box
+	this->bb.center = bb.center;
+	this->bb.radius = bb.radius;
+	
+	//Split datas
+	const std::size_t nbData = datas.size();
+	
+	DataContainer sDatas[8];
+	for(size_t i=0; i<8; ++i)
+		sDatas[i].reserve(nbData);
+		
+	for(auto&& d : datas)
+	{
+		//FIXME: Should be a generic conversion from DataPoint considering Data to XYZ
+		(sDatas[idx( TO_XYZ(pts,d) )]).emplace_back(d);
+	}
+	
+	for(size_t i=0; i<8; ++i)
+		sDatas[i].shrink_to_fit();
+	
+	//Compute new bounding boxes
+	BoundingBox boxes[8];
+	const T half_radius = this->bb.radius * 0.5;
+	for(size_t i=0; i<8; ++i)
+	{
+		const XYZ offset = offsetTable[i] * this->bb.radius;
+		boxes[i].radius = half_radius;
+		boxes[i].center = this->bb.center + offset;
+	}
+	
+	//For each child build recursively	
+	std::vector<std::thread> threads;
+	bool ret = true;
+	for(size_t i=0; i<8; ++i)
+	{
+		threads.push_back( std::thread( [maxSizeByNode, i, &pts, &sDatas, &boxes, this](){
```

Comment:
`std::async` with the `std::launch::async` policy (!) would have been better suited IMO (leaving more decisions to the implementation (e.g. actual number of threads to optimize hardware utilization).